webapp/src/components/backstage/playbook_runs/playbook_run_backstage/metrics/metrics_data.tsx

```diff
@@ -0,0 +1,148 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+import React, {forwardRef, useImperativeHandle, useRef, useState} from 'react';
+import {useIntl} from 'react-intl';
+
+import {VerticalSpacer} from 'src/components/backstage/playbook_runs/shared';
+
+import {RunMetricData} from 'src/types/playbook_run';
+import {Metric, MetricType} from 'src/types/playbook';
+import {ClockOutline, DollarSign, PoundSign} from '../../../playbook_edit/styles';
+import {isMetricValueValid, stringToMetric, metricToString} from '../../../playbook_edit/metrics/shared';
+
+import {useClickOutsideRef} from 'src/hooks';
+
+import MetricInput from './metric_input';
+
+interface MetricsProps {
+    metricsData: RunMetricData[];
+    metricsConfigs: Metric[];
+    isPublished: boolean;
+    onEdit: (metricsData: RunMetricData[]) => void;
+    flushChanges: () => void;
+}
+
+const MetricsData = forwardRef(({metricsData, metricsConfigs, isPublished, onEdit, flushChanges}: MetricsProps, ref) => {
+    const {formatMessage} = useIntl();
+
+    const initialValues = new Array(metricsConfigs.length);
+    metricsConfigs.forEach((mc, index) => {
+        const md = metricsData.find((metric) => metric.metric_config_id === mc.id);
+        initialValues[index] = md ? metricToString(md.value, mc.type) : '';
+    });
+    const [inputsValues, setInputsValues] = useState(initialValues);
```

Comment:
This is interesting. You could also use a function initializer, and create the initialValues inside that. This way the logic is only run once. Right now, it's run on every render.
So, something like:
```suggestion
    const [inputsValues, setInputsValues] = useState(() => {
        ...logic that returns an array of initial values
    });
```